cpp/include/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,489 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cassert>
+#include <functional>
+#include <iostream>
+#include <limits>
+
+namespace numeric {
+
+// This is a wrapper struct that enforces "strong typing"
+// at the construction site of the type. No implicit
+// conversions will be allowed and you will need to use the
+// name of the type alias (i.e. scale_type{0})
+template <typename T>
+struct weak_typedef {
+    T _t;
+    explicit weak_typedef(T t) : _t(t) {}
+    operator T() const { return _t; }
+};
+
+using scale_type = weak_typedef<int32_t>;
+
+enum class Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+template <typename T>
+constexpr inline auto is_supported_representation_type() {
+  return std::is_same<T, int32_t>::value
+      || std::is_same<T, int64_t>::value;
+}
+
+template <typename T>
+constexpr inline auto is_supported_construction_value_type() {
+  return std::is_integral      <T>::value
+      || std::is_floating_point<T>::value;
+}
+
+namespace detail {
+
+    // `exponent` comes from using scale_type = weak_typedef<int32_t>
+    //  we still need Rep to know what type we want for accumulator of ipow
+    template <typename Rep, Radix Base, typename T,
+              typename std::enable_if_t<(std::is_same<int32_t, T>::value
+                                      && is_supported_representation_type<Rep>())>* = nullptr>
+    Rep ipow(T exponent) {
+        // Integer Exponentiation by Squaring:
+        // https://simple.wikipedia.org/wiki/Exponentiation_by_squaring
+        // Note: this is the iterative equivalent of the recursive definition (faster)
+        // Quick-bench: http://quick-bench.com/Wg7o7HYQC9FW5M0CO0wQAjSwP_Y
```

Comment:
Are you concerned that quick-bench shows that your ipow is as fast as a no-op? Perhaps the benchmark is inaccurate.

Reply:
I was just trying to explain why I implemented it iteratively and not recursively.